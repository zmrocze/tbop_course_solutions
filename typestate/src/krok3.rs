pub struct Point {
  x: i32,
  y: i32,
  z: i32,
}

pub struct PointBuilder<X, Y, Z> {
  x: X,
  y: Y, 
  z: Z,
}

impl<> PointBuilder<(), (), ()> {
  pub fn new() -> Self {
    PointBuilder { x: (), y: (), z: () }
  }
}

impl<Y, Z> PointBuilder<(), Y, Z> {
  pub fn x(self, x: i32) -> PointBuilder<i32, Y, Z> {
    PointBuilder { x , y : self.y, z : self.z }
  }
}

impl<X, Z> PointBuilder<X, (), Z> {
  pub fn y(self, y: i32) -> PointBuilder<X, i32, Z> {
    PointBuilder { x : self.x , y : y, z : self.z }
  }
}

impl<X, Y> PointBuilder<X, Y, ()> {
  pub fn z(self, z: i32) -> PointBuilder<X, Y, i32> {
    PointBuilder { x :self.x, y : self.y, z }
  }
}

impl PointBuilder<i32, i32, i32> {
  pub fn build(self) -> Point {
    Point { x : self.x, y : self.y , z : self.z }
  }
}

#[test]
fn cant_compile() {
  let t = trybuild::TestCases::new();
  t.compile_fail("tests/cant_compile/*.rs");
  t.compile_fail("tests/cant_compile/*.rs");
}

#[test]
fn works() {
  let point2 = PointBuilder::new()
    .x(0)
    .y(1)
    .z(2)
    .build();

  let point1 = PointBuilder::new()
    .y(1)
    .x(0)
    .z(2)
    .build();
  println!("Hello, world! {} {}", point1.x, point2.y);
}

fn main () {

}