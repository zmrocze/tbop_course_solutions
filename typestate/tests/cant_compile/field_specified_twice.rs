use typestate::krok3::PointBuilder;

fn field_specified_twice() {
  let point = PointBuilder::new()
  .x(1)
  .x(1)
  .build();
}

fn main() {}