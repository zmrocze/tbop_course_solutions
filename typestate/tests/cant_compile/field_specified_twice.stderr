error[E0599]: no method named `x` found for struct `typestate::krok3::PointBuilder<i32, (), ()>` in the current scope
 --> tests/cant_compile/field_specified_twice.rs:6:4
  |
4 |     let point = PointBuilder::new()
  |  _______________-
5 | |   .x(1)
6 | |   .x(1)
  | |   -^
  | |   ||
  | |   |private field, not a method
  | |___|help: there is a method with a similar name: `y`
  |
